crates/telio-pmtu/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//1500 is the Ethernet MTU so we're not expecting PMTU to be higher than that.
+#[allow(unused)]
+const PMTU_RANGE: Range<u32> = 1000..1501;
+
+#[allow(unused)]
+const ICMP_ECHO_REQ_IDENT: u16 = 0x6C74; // "lt" (libtelio) in hex
+
+#[allow(unused)]
+const INTERVAL: Duration = Duration::from_secs(10 * 60); // 10 mins
+
+pub struct Entity {
+    #[allow(unused)]
+    socket_pool: Arc<telio_sockets::SocketPool>,
+    task: Option<JoinHandle<()>>,
+}
+
+impl Entity {
+    pub fn new(socket_pool: Arc<telio_sockets::SocketPool>) -> Self {
+        Self {
+            socket_pool,
+            task: None,
+        }
+    }
+
+    pub async fn run(&mut self, host: IpAddr) {
+        self.stop().await;
+        self.start(host);
+    }
+
+    pub async fn stop(&mut self) {
+        if let Some(task) = self.task.take() {
+            telio_utils::telio_log_debug!("Stopping PMTU discovery task");
+            task.abort();
+            let _ = task.await;
+        }
+    }
+
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    fn start(&mut self, host: IpAddr) {
+        telio_utils::telio_log_debug!("Starting PMTU discovery task");
+
+        use std::os::fd::AsRawFd;
+
+        use telio_lana::*;
+
+        async fn probe(
+            socket_pool: &telio_sockets::SocketPool,
+            host: IpAddr,
+        ) -> std::io::Result<u32> {
+            let sock = PMTUSocket::new(host)?;
+            telio_utils::telio_log_debug!("Making PMTU socket external");
+            socket_pool.make_external(sock.as_raw_fd());
+            sock.probe_pmtu().await
+        }
+
+        let pool = self.socket_pool.clone();
+        let task = tokio::spawn(async move {
+            let mut timer = tokio::time::interval(INTERVAL);
+
+            loop {
+                timer.tick().await;
+
+                match probe(&pool, host).await {
+                    Ok(pmtu) => {
+                        telio_utils::telio_log_info!("PMTU -> {host}: {pmtu}");
+
+                        let _ = lana! {
+                            send_developer_logging_log,
+                            pmtu as i32,
+                            telio_lana::moose::MeshnetappLogLevel::MeshnetappLogLevelInfo,
+                            format!("PMTU, OS: {}, ARCH: {}", std::env::consts::OS, std::env::consts::ARCH)
+                        };
+                    }
+                    Err(err) => telio_utils::telio_log_warn!("Failed to probe PMTU: {err:?}"),
+                }
+            }
+        });
+
+        self.task = Some(task);
+    }
+
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    fn start(&mut self, _: IpAddr) {}
+}
+
+impl Drop for Entity {
+    fn drop(&mut self) {
+        if let Some(task) = self.task.take() {
+            task.abort();
+        }
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+#[cfg(unix)]
+fn create_icmp_socket(host: &std::net::IpAddr) -> std::io::Result<tokio::net::UdpSocket> {
```

Comment:
Note that this socket does need to be "taken care of" in order to prevent it from going inside of a tunnel. 

* For linux, one needs to [setup fwmark](https://github.com/NordSecurity/libtelio/blob/v4.2.4/crates/telio-sockets/src/protector/linux.rs#L46)
* For android one needs to call [VPNService.protect()](https://developer.android.com/reference/android/net/VpnService#protect(int)) we already have a mechanism to do that via callbacks established.
* If other platforms would be needed, platform-specific mechanisms would be required too.

Keeping :arrow_up: in mind, I would recommend to move socket creation logic to `telio-sockets` module, as that module is responsible for exactly that, and have platform-dependant mechanisms (called "protectors") already implemented :).

Reply:
Oh wait, you are using `socket_pool.make_external(...)` which does mean, that all the required configuration for socket is done.

So this is purely structural thing, not functional, maybe that function would still belong close to `new_external_tcp_v4` or `new_internal_udp` or all other socket creation functions. As it is similar in purpose? What do you think?

Reply:
TBH I'd prefer keeping the PMTU module decoupled from the VPN itself. The primary reason is not to require PMTU nat-lab tests to set up the VPN things and the second one is to allow usage of PMTU for manual tests without the telio device configuration beforehand.
The `make_external` call does the job well IMHO, so all the requirements are ticked but the convenience and decoupling are preserved for the current implementation